ex_4/hashizume/EX4.py

```diff
@@ -0,0 +1,254 @@
+    for i in range(AC.shape[0] - 2):
+        if AC[i] < AC[i + 1] and AC[i + 1] > AC[i + 2]:
+            peak.append([i + 1, AC[i + 1]])
+    return np.array(peak)
+
+
+def calc_AC(x, win_length, hop_length, sr):
+    """
+    estimate f0 series
+    Calculate the autocorrelation for each frame
+    and extract the second peak
+
+    Parameters
+    ----------
+    x : input signal
+    win_length : length of window
+    hop_length : hop size
+    sr : sampling rate
+
+    returns
+    -------
+    f0 series of AC (sr / second peaks)
+
+    note
+    ----
+    The first peak is excluded,
+    so "the largest peak = the second peak"
+
+    """
+    f0series = []
+
+    # number of samples
+    N = x.shape[0]
+
+    # number of times to apply the window
+    T = int((N - hop_length) / hop_length)
+
+    for t in range(T):
+        # data cut out by the window
+        x_flame = x[t * hop_length : t * hop_length + win_length]
+
+        # calculate Autocorrelation
+        AC = AutoCorrelation(x_flame, x_flame.shape[0])
+
+        # narrow down peak candidates
+        peak = detect_peak(AC)
+        peak_index = peak[:, 0]
+        peak_value = peak[:, 1]
+
+        # sort in descending order of value
+        peak_value_sorted = np.sort(peak_value)[::-1]
+
+        second_peak_value = peak_value_sorted[0]
+
+        # index of "second peak"
+        second_peak_index = np.where(AC == second_peak_value)[0]
+
+        # calculate the f0 series
+        f0series.append(sr / second_peak_index[0])
+    return np.array(f0series)
+
+
+def db(x, dBref):
+    """
+    linear → db conversion
+
+    Parameters
+    ----------
+    x : input
+    dBref : decibel reference
+    """
+
+    y = 20 * np.log10(x / dBref)
+    return y
+
+
+def lin(x, dBref):
+    """
+    db → linear conversion
+
+    Parameters
+    ----------
+    x : input
+    dBref : decibel reference
+    """
+
+    y = dBref * 10 ** (x / 20)
+    return y
+
+
+def norm(spectrum, dBref):
+    """
+    normalization
+
+    Parameters
+    ----------
+    x : input
+    dBref : decibel reference
+    """
+
+    spectrum = lin(spectrum, dBref)
+    spectrum = np.abs(spectrum / (len(spectrum) / 2))
+    spectrum = db(spectrum, dBref)
+    return spectrum
+
+
+def levinson(r):
+    """
+    levinson durbin algorithm
+    """
+    a = np.zeros(len(r))
+    a[0] = 1
+    sigma = r[0]
+    for p in range(1, a.shape[0]):
+        w = np.sum(a[:p] * r[p:0:-1], axis=0)
+        k = w / sigma
+        sigma = sigma - k * w
+
+        a[1 : p + 1] = a[1 : p + 1] - k * a[p - 1 :: -1]
+    e = np.sqrt(sigma)
+    return a, e
+
+
+def lpc(x, win_length, deg):
+    ac = AutoCorrelation(x, x.shape[0])
+    r = ac[:deg]
+    a, e = levinson(r)
+    env = e / np.fft.rfft(a, win_length, axis=0)
+    env = librosa.amplitude_to_db(np.abs(env))
+
+    return env
+
+
+def main():
+    args = sys.argv
+    wav_filename = args[1]
+    origin_signal, sr = librosa.load(wav_filename)
```

Comment:
これは直接の原因ではないかもしれませんが, librosa.load はsrが与えられないとデフォルト値22.5kHzにリサンプリングするらしいです. 元ファイルのサンプリングレートで読み込むには None を渡します.
```suggestion
    origin_signal, sr = librosa.load(wav_filename, sr=None))
```
こうした方がきれいな結果が出ました.

Reply:
@AtsushiMiyashita317 @zume06 
ちなみに、あらかじめサンプリングレートが分かっていないファイルを使う場合は
```
import soundfile as sf
x, sr = sf.read(path)
```
とやると、収録されたサンプリングレートで配列に格納して、どのサンプリングレートだったかもわかるので便利です